### PR TITLE
feat(): add typings to the blueprints.

### PR DESCRIPTION
**Pull request by hansl**
Simply add typings to the blueprints, and use the browser ones in `tsconfig`.

This should get the upgrade to angular beta.2 unblocked.

### CODE REVIEW

File: addon/ng2/blueprints/ng2/files/src/tsconfig.json [hunk 1]
```diff
@@ -8,9 +8,14 @@
...
     "noEmitOnError": true,
     "noImplicitAny": false,
+    "outDir": "../dist/",
     "rootDir": ".",
     "sourceMap": true,
     "sourceRoot": "/",
     "target": "es5"
-  }
+  },
+  "filesGlob": [
```
Comment (alexeagle) [inline addon/ng2/blueprints/ng2/files/src/tsconfig.json]: this is not nice. Only atom understands this property, not tsc.

Your options are
* use exclude, meaning all `.ts` and `.d.ts` files in this directory and children
* use files =[] which means you need to specify the application entry point(s) as well as the typings/browser.d.ts file (or the typings/main.d.ts file if targetting node runtime)

File: addon/ng2/blueprints/ng2/files/package.json [hunk 2]
```diff
@@ -27,6 +31,7 @@
     "karma-chrome-launcher": "^0.2.1",
     "karma-jasmine": "^0.3.6",
     "protractor": "^3.0.0",
-    "typescript": "^1.7.3"
+    "typescript": "^1.7.3",
```
Comment (alexeagle) [inline addon/ng2/blueprints/ng2/files/package.json]: maybe we should opt angular-cli users into the TS 1.8 beta. Then you get features like comments in the tsconfig.json which is really nice in a templated file - gives the user in-situ instructions for editing it

File: addon/ng2/blueprints/ng2/files/package.json [hunk 3]
```diff
@@ -8,6 +8,10 @@
       "playSound": false
     }
   },
+  "scripts": {
+    "start": "ng server",
+    "postinstall": "typings install --ambient"
```
Comment (alexeagle) [inline addon/ng2/blueprints/ng2/files/package.json]: you should test what happens when a typing is removed from typings.json - make sure that this will remove it from the typings dir. otherwise contributors end up with stale extra typings files in their working dir

File: addon/ng2/blueprints/ng2/files/src/tsconfig.json (hunk 1 above)
Comment (hansl) [inline addon/ng2/blueprints/ng2/files/src/tsconfig.json]: Done, replaced with an exclude for the main file (we only need browsers AFAIK).

File: addon/ng2/blueprints/ng2/files/package.json (hunk 2 above)
Comment (hansl) [inline addon/ng2/blueprints/ng2/files/package.json]: We should make those kind of version upgrade a separate PR, but I agree.

File: addon/ng2/blueprints/ng2/files/package.json (hunk 3 above)
Comment (hansl) [inline addon/ng2/blueprints/ng2/files/package.json]: Does not remove the typings removed from typings.json.

I filed https://github.com/typings/typings/issues/170 to do just that.

**Comment by alexeagle** [comment]
LGTM

File: addon/ng2/blueprints/ng2/files/src/tsconfig.json [hunk 4]
```diff
@@ -8,9 +8,13 @@
...
     "noImplicitAny": false,
+    "outDir": "../dist/",
     "rootDir": ".",
     "sourceMap": true,
     "sourceRoot": "/",
     "target": "es5"
-  }
+  },
+  "exclude": [
+    "../typings/main.d.ts"
```
Comment (filipesilva) [inline addon/ng2/blueprints/ng2/files/src/tsconfig.json]: Excluding the file means the typings we have will never be accessible.

This is the problem I ran into: typescript allows you to either list all files, or exclude some of them. To use exclude you'd need typings to be installed on the `src/` folder, and exclude the ones you don't want.

Comment (filipesilva) [inline addon/ng2/blueprints/ng2/files/src/tsconfig.json]: On the other hand, beta.2 breaks `ng serve` for the same reason, so... a bit of a catch-22 here. You can test this by just upgrading npm angular2 on the test project.

Comment (hansl) [inline addon/ng2/blueprints/ng2/files/src/tsconfig.json]: We cannot use `files` because every unittest is its own entry point.

**Comment by hansl** [comment]
I'm going to try to upgrade my local app to beta.2 and see what happens. BRB.

**Comment by hansl** [comment]
@alexeagle  @filipesilva  @IgorMinar 

Adding these typings work, when we pass in `rootFilePaths` to the `broccoli-typescript`. It's a custom property that is used by the TypeScript language service, defined in the ServiceHost. Basically we append our list of TS files to that property. `rootFilePaths` isn't standard though, in the official TypeScript host it's named `getScriptFileNames`.

After looking deeply into this (and it also gazed back info me, Nietzsche-style), I see a few problems with our current scripts in `angular-cli`:

* nothing is typescript, but the source of the files is definitely TypeScript. We should replace our JS files by their TS sources and output the proper JS on prepublish. That's not much work I'll see if I can fit this into my schedule.
* moving the `typings.json` to `src/` isn't as convenient as it looks. It removes the use of `process.cwd()` from the angular broccoli class, but you need to `cd src` everytime you use `typings`, which isn't great.

Finally, I added a move to beta.3 in here, but I will remove it before I submit. The reason is we need this change in our npm package before we upgrade to beta.3. So unless you guys tell me it's okay to have a half-release time where apps generated from master do not work (because the npm package isn't good), then I'll undo the upgrade. @filipesilva will be happy to know this is super straightforward though; just change the scaffold dependency up to beta.3 and the zone to .11 and it's good to go as is.

If you want to use this PR as a local test; once the app is generated you need to replace the `node_modules/` version of `angular-cli` with your local copy. Basically, do the following shell commands:

```sh
./angular-cli/bin/ng new foo
cp -rf ./angular-cli foo/node_modules/
cd foo/
ng build  # will work
```

Please note that the `tsc` will not work unless you add all the entry points to the `tsconfig.json` files key. The problem is each spec file is its own entry point, so this would become unmanageable.

With that, I'm taking a break. Please take a look at your convenience.

**Comment by filipesilva** [comment]
Whoa @hansl , great work! I can't follow up on this until next wednesday as I'm just leaving for a vacations, but there's enough info here for someone else to merge for now.

Again, thanks for looking into this, it was driving me nuts.

**Comment by filipesilva** [comment]
lgtm :+1:
